x-pack/plugins/upgrade_assistant/server/routes/node_disk_space.ts

```diff
@@ -0,0 +1,152 @@
+ * 2.0.
+ */
+
+import type { ClusterGetSettingsResponse } from '@elastic/elasticsearch/api/types';
+import { ByteSizeValue } from '@kbn/config-schema';
+import { API_BASE_PATH } from '../../common/constants';
+import { versionCheckHandlerWrapper } from '../lib/es_version_precheck';
+import { RouteDependencies } from '../types';
+
+interface NodeWithLowDiskSpace {
+  nodeId: string;
+  nodeName: string;
+  used: string;
+  lowDiskWatermarkSetting: string;
+}
+
+// Buffer percentage used to determine if the user has low disk space
+// For example, if a user has used 80%, and has defined a low disk watermark setting of 85%, we would warn the user
+const LOW_DISK_SPACE_BUFFER_PERCENTAGE = 5;
+
+const getLowDiskWatermarkSetting = (clusterSettings: ClusterGetSettingsResponse) => {
+  const { defaults, persistent, transient } = clusterSettings;
+
+  const defaultLowDiskWatermarkSetting =
+    defaults && defaults['cluster.routing.allocation.disk.watermark.low'];
+  const transientLowDiskWatermarkSetting =
+    transient && transient['cluster.routing.allocation.disk.watermark.low'];
+  const persistentLowDiskWatermarkSetting =
+    persistent && persistent['cluster.routing.allocation.disk.watermark.low'];
+
+  // ES applies cluster settings in the following order of precendence: transient, persistent, default
+  if (transientLowDiskWatermarkSetting) {
+    return transientLowDiskWatermarkSetting;
+  } else if (persistentLowDiskWatermarkSetting) {
+    return persistentLowDiskWatermarkSetting;
+  } else if (defaultLowDiskWatermarkSetting) {
+    return defaultLowDiskWatermarkSetting;
+  }
+
+  // May be undefined if defined in elasticsearch.yml
+  return undefined;
+};
+
+export function registerNodeDiskSpaceRoute({ router, lib: { handleEsError } }: RouteDependencies) {
+  router.get(
+    {
+      path: `${API_BASE_PATH}/node_disk_space`,
+      validate: false,
+    },
+    versionCheckHandlerWrapper(
+      async (
+        {
+          core: {
+            elasticsearch: { client },
+          },
+        },
+        request,
+        response
+      ) => {
+        try {
+          const { body: clusterSettings } = await client.asCurrentUser.cluster.getSettings({
+            flat_settings: true,
+            include_defaults: true,
+          });
+
+          const lowDiskWatermarkSetting = getLowDiskWatermarkSetting(clusterSettings);
+
+          if (lowDiskWatermarkSetting) {
+            const { body: nodeStats } = await client.asCurrentUser.nodes.stats({
+              metric: 'fs',
+            });
+
+            const nodeIds = Object.keys(nodeStats.nodes);
+
+            const nodesWithLowDiskSpace: NodeWithLowDiskSpace[] = [];
+
+            nodeIds.forEach((nodeId) => {
+              const node = nodeStats.nodes[nodeId];
+              const byteStats = node.fs.total;
+              const { total_in_bytes: totalInBytes, available_in_bytes: availableInBytes } =
```

Comment:
@masseyke can you verify that it's the correct approach to use `available_in_bytes` (vs. `free_in_bytes`) from the node stats API?

Reply:
Current approach based on this finding: https://stackoverflow.com/questions/16302348/whats-the-difference-between-getusablespace-and-getunallocatedspace-of-filestor#comment111419227_31435262

Reply:
Yeah `available_in_bytes` is the one that makes more sense to me to use here.